src/app/donation-start/donation-start-form/donation-tipping-slider/donation-tipping-slider.component.ts

```diff
@@ -0,0 +1,214 @@
+        this.isMoving = true;
+        this.move(e);
+      });
+
+      this.bar.nativeElement.addEventListener('touchend', () => {
+        this.isMoving = false;
+      });
+    }
+  };
+
+  // detect changes in the donationAmount input
+  ngOnChanges(changed: any) {
+    this.max = this.bar.nativeElement.offsetWidth - this.handle.nativeElement.offsetWidth;
+    if (changed.donationAmount.currentValue != changed.donationAmount.previousValue) {
+      this.calcAndSetPercentage();
+      this.calcAndSetTipAmount();
+      this.adjustDonationPercentageAndValue();
+      this.updateHandlePosition(undefined);
+      this.onHandleMoved(this.derivedPercentage, this.tipAmount);
+    }
+  }
+
+  ngOnDestroy() {
+    this.documentMousemoveUnlistener();
+    this.documentMouseupUnlistener();
+    this.documentTouchmoveUnlistener();
+  }
+
+  calcAndSetPercentage() {
+  // the calculation results from mouse click
+   if (this.isMoving) {
+    this.derivedPercentage = Math.round((this.position / this.max) * this.percentageEnd >= 1 ? (this.position / this.max) * this.percentageEnd : 1);
+   }
+  // the calculation results from input changes
+   else if (!this.disableDefaults){
+      if (!this.donationAmount) {
+        this.derivedPercentage = 0;
+      }
+      if (this.donationAmount >= 1000) {
+        this.derivedPercentage = 7.5;
+      } else if (this.donationAmount >= 300) {
+        this.derivedPercentage = 10;
+      } else if (this.donationAmount >= 100) {
+        this.derivedPercentage = 12.5;
+      } else {
+        this.derivedPercentage = 15;
+      }
+    }
+  }
+
+  format(currencyCode: 'GBP' | 'USD', amount: number) {
+    return Intl.NumberFormat('en-GB', {
+      style: 'currency',
+      currency: currencyCode,
+    }).format(amount);
+  }
+
+  move = (e: MouseEvent | TouchEvent) => {
+    if (this.isMoving) {
+      this.disableDefaults = true;
+      this.max = this.bar.nativeElement.offsetWidth - this.handle.nativeElement.offsetWidth;
+
+      if (window.TouchEvent && e instanceof TouchEvent) {
+        this.pageX = e.touches[0]?.pageX;
+      } else {
+        // we Know e is a MouseEvent because all platforms that supports TouchEvent would also have
+        // a truthy window.TouchEvent - see https://stackoverflow.com/a/32882849/2803757
+        this.pageX = (e as MouseEvent).pageX;
+      }
+
+      if (this.pageX !== undefined) {
+        if(this.derivedPercentage) {
+          this.calcAndSetTipAmount();
+          this.updateHandlePosition(e);
+          this.adjustDonationPercentageAndValue();
+          this.onHandleMoved(this.derivedPercentage, this.tipAmount);
+        }
+      }
+    }
+
+  }
+
+  calcAndSetTipAmount() {
+    if (this.donationAmount < 55) {
+      this.tipAmount = this.donationAmount * (this.derivedPercentage / 100);
+    } else {
+      this.tipAmount = Math.round(this.donationAmount * (this.derivedPercentage / 100));
```

Comment:
maybe getting nitpicky but better not to repeat the calculation code? Could just calculate it outside the conditional and then conditionally round?